File: sgilt-core/src/main/java/net/franzka/sgilt/core/reservation/mapper/ReservationMapper.java
package net.franzka.sgilt.core.reservation.mapper;

import net.franzka.sgilt.core.reservation.domain.Reservation;
import net.franzka.sgilt.core.reservation.domain.ReservationStatus;
import net.franzka.sgilt.core.reservation.dto.ActiveReservationItemDto;
import net.franzka.sgilt.core.reservation.dto.ProReservationDetailDto;
import net.franzka.sgilt.core.reservation.dto.ProReservationSummaryDto;
import net.franzka.sgilt.core.reservation.dto.ReservationMetaDto;
import net.franzka.sgilt.core.reservation.dto.ReservationSummaryDto;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import org.mapstruct.Named;
import org.mapstruct.ValueMapping;

@Mapper(componentModel = "spring")
public interface ReservationMapper {

    @ValueMapping(source = "NEW",                                    target = "nouvelle")
    @ValueMapping(source = "IN_DISCUSSION",                          target = "en_discussion")
    @ValueMapping(source = "CONFIRMED",                              target = "confirmee")
    @ValueMapping(source = "DONE",                                   target = "realisee")
    @ValueMapping(source = "REFUSED_PRE_CONTACT",                    target = "refusee")
    @ValueMapping(source = "REFUSED_POST_CONTACT",                   target = "refusee")
    @ValueMapping(source = "CANCELED_BY_CLIENT_PRE_CONTACT",         target = "annulee")
    @ValueMapping(source = "CANCELED_BY_CLIENT_POST_CONTACT",        target = "annulee")
    @ValueMapping(source = "CANCELED_POST_CONFIRMATION",             target = "annulee")
    String mapStatus(ReservationStatus status);

    @Mapping(source = "prestataire.id",          target = "prestataireId")
    @Mapping(source = "prestataire.name",        target = "prestataireName")
    @Mapping(source = "prestataire.avatar",      target = "prestatairePhoto")
    @Mapping(source = "prestataire.categoryKey", target = "category")
    @Mapping(target = "unreadNotesCount",        constant = "0")
    ReservationSummaryDto toSummaryDto(Reservation reservation);

    @Mapping(source = "id",                                        target = "reservationId")
    @Mapping(source = "evenement.id",                              target = "evenementId")
    @Mapping(source = "evenement.title",                           target = "evenementTitle")
    @Mapping(source = "prestataire.slug",                          target = "prestataireSlug")
    @Mapping(source = "prestataire.name",                          target = "prestataireName")
    @Mapping(source = "prestataire",                               target = "prestataireAvatar", qualifiedByName = "resolveAvatar")
    @Mapping(source = "status",                                    target = "status")
    ActiveReservationItemDto toActiveItemDto(Reservation reservation);

    @Mapping(source = "evenement.title",     target = "evenementTitre")
    @Mapping(source = "evenement.eventType", target = "evenementType")
    @Mapping(source = "evenement.imagePath", target = "image")
    @Mapping(source = "date",                target = "datePrestation")
    @Mapping(source = "status",              target = "statut")
    @Mapping(target = "unreadNotesCount",    constant = "0")
    ProReservationSummaryDto toProReservationSummaryDto(Reservation reservation);

    @Mapping(source = "status",                   target = "statut")
    @Mapping(source = "prestataire.categoryKey",  target = "category")
    @Mapping(source = "prestataire.name",         target = "prestataireName")
    @Mapping(source = "prestataire",              target = "prestatairePhoto", qualifiedByName = "resolveAvatar")
    @Mapping(source = "evenement.title",          target = "evenementTitre")
    @Mapping(source = "evenement.eventType",      target = "evenementType")
    @Mapping(source = "evenement.imagePath",      target = "evenementImagePath")
    @Mapping(source = "evenement.date",           target = "evenementDate")
    @Mapping(source = "evenement.ville",          target = "evenementVille")
    @Mapping(source = "utilisateur.firstName",    target = "clientFirstName")
    @Mapping(source = "utilisateur.lastName",     target = "clientLastName")
    @Mapping(source = "utilisateur.phone",        target = "clientPhone")
    @Mapping(source = "utilisateur.email",        target = "clientEmail")
    ProReservationDetailDto toProReservationDetailDto(Reservation reservation);

    @Mapping(source = "prestataire.id",          target = "prestataireId")
    @Mapping(source = "prestataire.name",        target = "prestataireName")
    @Mapping(source = "prestataire",             target = "prestatairePhoto", qualifiedByName = "resolveAvatar")
    @Mapping(source = "prestataire.categoryKey", target = "category")
    @Mapping(source = "status",                  target = "status")
    @Mapping(target = "unreadNotesCount",        constant = "0")
    ReservationMetaDto toReservationMetaDto(Reservation reservation);

    @Named("resolveAvatar")
    default String resolveAvatar(net.franzka.sgilt.core.prestataire.domain.Prestataire prestataire) {
        return prestataire.getAvatar() != null ? prestataire.getAvatar() : prestataire.getHeroImage();
    }
}

File: sgilt-front/app/data/reservation/mapper/reservationMapper.ts
import type { ReservationMetaDto } from '../dto/ReservationMetaDto'
import type { ProReservationSummaryDto } from '../dto/ProReservationSummaryDto'
import type { ProReservationDetailDto } from '../dto/ProReservationDetailDto'
import type { ActiveReservationItemDto, ActiveReservationsDto } from '../dto/ActiveReservationsDto'
import type { ReservationMeta } from '../domain/ReservationMeta'
import type { ProReservationSummary } from '../domain/ProReservationSummary'
import type { ProReservationDetail } from '../domain/ProReservationDetail'
import type { ActiveReservations, ActiveReservationItem } from '../domain/ActiveReservation'
import type { ReservationStatus } from '../domain/ReservationStatus'
import type { EventDetail } from '~/data/evenement/domain/EventDetail'
import type { ClientContactInfo } from '../domain/ClientContactInfo'

export function mapReservationMeta(dto: ReservationMetaDto): ReservationMeta {
  return {
    id: dto.id,
    prestataireId: dto.prestataireId,
    prestataireName: dto.prestataireName,
    prestatairePhoto: dto.prestatairePhoto ?? undefined,
    category: dto.category,
    status: dto.status as ReservationMeta['status'],
    unreadNotesCount: dto.unreadNotesCount,
  }
}

export function mapProReservationSummary(dto: ProReservationSummaryDto): ProReservationSummary {
  return {
    id: dto.id,
    titre: dto.evenementTitre,
    eventType: dto.evenementType ?? undefined,
    date: new Date(dto.datePrestation),
    statut: dto.statut as ReservationStatus,
    image: dto.image ?? undefined,
    unreadNotesCount: dto.unreadNotesCount,
  }
}

function mapProReservationDetailEvent(reservationId: string, dto: ProReservationDetailDto): EventDetail {
  return {
    id: reservationId,
    title: dto.evenementTitre,
    date: dto.evenementDate ? new Date(dto.evenementDate) : undefined,
    eventType: dto.evenementType ?? undefined,
    ville: dto.evenementVille ?? undefined,
    coverImage: dto.evenementImagePath ?? null,
    sharedNote: '',
    reservations: [],
    journal: [],
    mood: 'defaut',
    countdown: 'serein',
  }
}

function mapProReservationDetailClientInfo(dto: ProReservationDetailDto): ClientContactInfo {
  return {
    firstName: dto.clientFirstName,
    lastName: dto.clientLastName ?? undefined,
    phone: dto.clientPhone ?? '',
    email: dto.clientEmail,
  }
}

export function mapProReservationDetail(reservationId: string, dto: ProReservationDetailDto): ProReservationDetail {
  return {
    id: dto.id,
    prestataireId: '',
    prestataireName: dto.prestataireName,
    prestatairePhoto: dto.prestatairePhoto ?? undefined,
    category: dto.category,
    status: dto.statut as ProReservationDetail['status'],
    unreadNotesCount: 0,
    event: mapProReservationDetailEvent(reservationId, dto),
    clientInfo: mapProReservationDetailClientInfo(dto),
    progressType: null,
    progressValue: null,
    phraseInfoState: null,
  }
}

function mapActiveReservationItem(dto: ActiveReservationItemDto): ActiveReservationItem {
  return {
    reservationId: dto.reservationId,
    status: dto.status as ReservationStatus,
    evenementId: dto.evenementId,
    evenementTitle: dto.evenementTitle,
    prestataireSlug: dto.prestataireSlug,
    prestataireName: dto.prestataireName,
    prestataireAvatar: dto.prestataireAvatar,
  }
}

export function mapActiveReservations(dto: ActiveReservationsDto): ActiveReservations {
  return {
    items: dto.items.map(mapActiveReservationItem),
    hasConfirmed: dto.hasConfirmed,
  }
}
